### DISCUSSION TITLE
mediaDevices and webrtc

### DISCUSSION THREAD

**Discussion by oxzjh** [wailsapp/wails#2697] [Q&A]
### Describe the solution you'd like

Not support mediaDevices and webrtc on Linux

**Comment by torabian** [comment]
navigator.mediaDevices is undefined, how can we record user audio on wails?

**Comment by oxzjh** [comment]
Mediadevices and WebRTC not supported on Linux.
Dose Wails will embed chromium instead of webkit to support advance features

**Reply by learncodingforweb** [reply]
I am facing same problem, not able to open usb camera using wails running on linux? any update?

```
# Wails

Version         | v2.5.1
Package Manager | apt   

# Dependencies

Dependency | Package Name          | Status    | Version                
*docker    | docker.io             | Installed | 24.0.2                 
gcc        | build-essential       | Installed | 12.9ubuntu3            
libgtk-3   | libgtk-3-dev          | Installed | 3.24.33-1ubuntu2       
libwebkit  | libwebkit2gtk-4.0-dev | Installed | 2.38.6-0ubuntu0.22.04.1
npm        | npm                   | Installed | 9.5.0                  
*nsis      | nsis                  | Installed | v3.08-2                
pkg-config | pkg-config            | Installed | 0.29.2-1ubuntu3        
* - Optional Dependency

```

**Reply by zhaojunliing** [reply]
Did you solve it?

**Comment by learncodingforweb** [comment]
any update?

**Comment by marcio199226** [comment]
any update ? I have the same issue on macOS but the strange part that I'm pretty sure that it works 1 month ago....

**Reply by leaanthony** [reply]
Hi! It looks like you're asking about browser API support. Wails does not embed a browser, but a webview, so all the browser APIs may not be available. Here are some links that may help you find what you are looking for. Let us know how you get on!

Windows:
- [Webview2 Documentation](https://learn.microsoft.com/en-us/microsoft-edge/webview2/)
- [Webview2 Release Notes](https://learn.microsoft.com/en-us/microsoft-edge/webview2/release-notes?tabs=win32cpp)
- [Webview2 Issue tracker](https://github.com/MicrosoftEdge/WebView2Feedback/issues)

Mac:
- [WkWebView](https://developer.apple.com/documentation/webkit/wkwebview)

Linux:
- [GtkWebkit](https://webkitgtk.org/)
- [Documentation](https://webkitgtk.org/reference/webkit2gtk/stable/index.html)
- [Release Notes](https://planet.webkitgtk.org/)
- [Issue Tracker Search](https://bugs.webkit.org/query.cgi)

**Comment by fairking** [comment]
> Dose Wails will embed chromium instead of webkit to support advance features

IMHO We should not use webview to access any devices. It must be done in backend instead. I don't like the idea to use chromium, which is a browser, and which is not what webview is for. WebView must be minimal and serve as a UI renderer. We have a backend for that and we can do everything we want when it comes to hardware integration.
See:
- #3225 
- #3212 
- #3168 
- #883

**Reply by leaanthony** [reply]
I agree with the sentiment. We have no plans to embed a browser. The issue is that streaming from Go to webview is not fully supported by the webviews yet so it's very hard to actually achieve the equivalent functionality.

**Reply by ccutch** [reply]
I was really excited to use wails, but it seems a little silly to use a framework that can't actually integrate with the hardware on the machine.

**Reply by torabian** [reply]
Wails is not aiming to support everything. I had integrated it with camera and microphone on mac at least, with a little bit more digging it would be possilbe.

**Reply by ccutch** [reply]
I understand that it is not trying to support everything, and that with effort this problem can be solved. But I can't care more about a project than the core contributors do. Right?

**Reply by ccutch** [reply]
I will use it again, and still suggest it as the onboarding curve is the best. But finding a ceiling that has been hit many times before is frustrating.

**Reply by fairking** [reply]
1. I think WebView name is not appropriate when it comes to the HtmlRenderer for Wails.
WebView is used to show pages from the web. While in Wails we use HTML as UI presentation, same as QT is using QML for UI.
2. I don't think HTTP communication is appropriate for backend communication. HTTP used to communicate between Server and Client remotely. What is the reason to use Wails, when I can simple build a normal web app and use it locally with browser. HTTP server requires a lot of resources and system permissions. The Wails app must communicate between WebView and backend natively. Otherwise again what is the reason to use Wails, it does not make any sense.
3. When it comes to device integration, again, it is not UI job to do that. It is a backend job to save files, talk to database, stream camera or RS and Serial Port communication or even access web APIs.

In my opinion Wails and Tauri and maybe other Frameworks must come all together and write it's own Html renderer instead. And that's I think what Tauri is [trying to do](https://twitter.com/TauriApps/status/1505618001126731781?). So I would consider Wails to join that team.

BTW. HTML5 is the best UI language I ever used. Which can be used in any web or desktop app.

References: 
- https://github.com/malisipi/mui/issues/1
- https://github.com/webview/webview/issues/851
- https://github.com/webview/webview/issues/1068

**Comment by guoqiang1** [comment]
How to get microphone input ?

**Reply by fairking** [reply]
You should search for any go packages which can do that. Make sure it supports multiple platforms.
I don't think you should use WebView for that (IMHO), see the previous discussion.

https://pkg.go.dev/search?q=microphone
